codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/XmlShapeDeserVisitor.java

```diff
@@ -0,0 +1,207 @@
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.typescript.codegen.integration.DocumentMemberDeserVisitor;
+import software.amazon.smithy.typescript.codegen.integration.DocumentShapeDeserVisitor;
+import software.amazon.smithy.typescript.codegen.integration.ProtocolGenerator.GenerationContext;
+
+/**
+ * Visitor to generate deserialization functions for shapes in XML-document
+ * based document bodies.
+ *
+ * No standard visitation methods are overridden; function body generation for all
+ * expected deserializers is handled by this class.
+ *
+ * Timestamps are deserialized from {@link Format}.DATE_TIME by default.
+ *
+ * @see <a href="https://awslabs.github.io/smithy/spec/xml.html">Smithy XML traits.</a>
+ */
+final class XmlShapeDeserVisitor extends DocumentShapeDeserVisitor {
+
+    XmlShapeDeserVisitor(GenerationContext context) {
+        super(context);
+    }
+
+    private DocumentMemberDeserVisitor getMemberVisitor(String dataSource) {
+        return new XmlMemberDeserVisitor(getContext(), dataSource, Format.DATE_TIME);
+    }
+
+    @Override
+    protected void deserializeCollection(GenerationContext context, CollectionShape shape) {
+        TypeScriptWriter writer = context.getWriter();
+        Shape target = context.getModel().expectShape(shape.getMember().getTarget());
+
+        // Dispatch to the output value provider for any additional handling.
+        writer.openBlock("return (output || []).map((entry: any) =>", ");", () -> {
+            writer.write(target.accept(getMemberVisitor("entry")));
+        });
+    }
+
+    @Override
+    protected void deserializeDocument(GenerationContext context, DocumentShape shape) {
+        throw new CodegenException(String.format("Cannot deserialize Document types on AWS XML protocols, shape: %s.",
+                shape.getId()));
+    }
+
+    @Override
+    protected void deserializeMap(GenerationContext context, MapShape shape) {
+        TypeScriptWriter writer = context.getWriter();
+        Shape target = context.getModel().expectShape(shape.getValue().getTarget());
+
+        // Get the right serialization for each entry in the map. Undefined
+        // outputs won't have this deserializer invoked.
+        writer.write("let mapParams: any = {};");
+        writer.openBlock("Object.keys(output).forEach(key => {", "});", () -> {
+            // Dispatch to the output value provider for any additional handling.
+            writer.write("mapParams[key] = $L;", target.accept(getMemberVisitor("output[key]")));
+        });
+        writer.write("return mapParams;");
+    }
+
+    @Override
+    protected void deserializeStructure(GenerationContext context, StructureShape shape) {
+        TypeScriptWriter writer = context.getWriter();
+
+        // Prepare the document contents structure.
+        // Use a TreeMap to sort the members.
+        Map<String, MemberShape> members = new TreeMap<>(shape.getAllMembers());
```

Comment:
This is already a TreeMap: https://github.com/awslabs/smithy/blob/master/smithy-model/src/main/java/software/amazon/smithy/model/shapes/NamedMembersShape.java#L41